packages/pwa-kit-runtime/src/ssr/server/express.js

```diff
@@ -460,3 +463,102 @@ export const getRuntime = () => {
 
     return boundRuntime
 }
+
+/**
+ * Patches `res.setHeader` to ensure that the Content-Security-Policy header always includes the
+ * directives required for PWA Kit to work.
+ * @param {express.Request} req Express request object
+ * @param {express.Response} res Express response object
+ * @param {express.NextFunction} next Express next callback
+ */
+export const defaultSecurityHeaders = (req, res, next) => {
+    /** CSP-compatible origin for Runtime Admin. */
+    // localhost doesn't include a protocol because different browsers behave differently :\
+    const runtimeAdmin = isRemote() ? 'https://runtime.commercecloud.com' : 'localhost:*'
+    /**
+     * Map of directive names/values that are required for PWA Kit to work. Array values will be
+     * merged with user-provided values; boolean values will replace user-provided values.
+     * @type Object.<string, string[] | boolean>
+     */
+    const directives = {
+        'connect-src': ["'self'", runtimeAdmin],
+        'frame-ancestors': [runtimeAdmin],
+        'img-src': ["'self'", 'data:'],
+        'script-src': ["'self'", "'unsafe-eval'", runtimeAdmin],
+        // Always upgrade insecure requests when deployed, never upgrade on local dev server
+        'upgrade-insecure-requests': isRemote()
+    }
+
+    const setHeader = res.setHeader
+    res.setHeader = (name, value) => {
```

Comment:
We do have an existing place where we apply patches:

https://github.com/SalesforceCommerceCloud/pwa-kit/blob/a726c4974490d23ed4869223a038ad7690847f8b/packages/pwa-kit-runtime/src/ssr/server/build-remote-server.js#L1134-L1193

I don't have a strong opinion about whether this patch should also be applied there.

Reply:
Seems like it would be a bit awkward to apply the patch there, as that's always run, and this middleware is conditional.

Reply:
That's fair